### ISSUE TITLE
Tensorflow Model Deployment Type Errors

### ISSUE THREAD

**Issue by lukealexmiller** [vespa-engine/vespa#10709]
I have an issue regarding loading of Tensorflow models (trained using Keras API) which seems to be related to some of the tensor reshaping discussed in my previous issue (https://github.com/vespa-engine/vespa/issues/9662) - thanks for the great feedback on this!

Using Vespa 7.74.17, I have been able to load models similar to those mentioned in the previous issue, which were trained in Keras, and have confirmed inference results between TF and Vespa. 

However, to deploy these models in the application package it was necessary to ‘freeze’ and ‘optimize’ the models similar to the example shown here (https://medium.com/google-cloud/optimizing-tensorflow-models-for-serving-959080e9ddbf) in order to eliminate the VarHandleOp/ReadVariableOp/other ResourceVariable references. 
Please see example at Model_A/optimized_v2/savedmodel.pbtxt: https://drive.google.com/open?id=1lghvIRH5GAN3vDvzLjydeCFkuQ1hmLw7

The relevant snippet of the searchdocument is:   
```
 field embedding type tensor(d0[1],d1[32]) {
      indexing: attribute | summary
    }

  rank-profile mf_rank {
    function input_embedding_product_slug() {
      expression: attribute(embedding)
    }
    function input_embedding_user_guid() {
      expression: query(user_embedding)
    }
    first-phase {
      expression: tensorflow("factorization_machine_v1", "serving_default", "output")
    }
  }
```

Using the same architecture, aside from some cosmetic changes to my codebase, I have trained another model (see Model_B/factorization_machine_v1/optimized/savedmodel.pbtxt https://drive.google.com/open?id=1ZJ_VMnfl991Jnh4dH7rkruGnJ41tCyHv), which fails on deployment with the following error:

```
ERROR: rank expression contains type errors: join(rename(map(join(reduce(join(join(reduce(join(join(rankingExpression(imported_ml_function_factorization_machine_v1_input_embedding_product_slug), reduce(map(join(join(reduce(map(rankingExpression(imported_ml_function_factorization_machine_v1_input_embedding_product_slug), f(a)(a * a)), sum, d0), tensor(d0[1])(1.0), f(a,b)(a * b)), 9.999999960041972E-13, f(a,b)(max(a,b))), f(a)(1.0 / sqrt(a))), sum, d0), f(a,b)(a * b)), join(rankingExpression(imported_ml_function_factorization_machine_v1_input_embedding_user_guid), reduce(map(join(join(reduce(map(rankingExpression(imported_ml_function_factorization_machine_v1_input_embedding_user_guid), f(a)(a * a)), sum, d0), tensor(d0[1])(1.0), f(a,b)(a * b)), 9.999999960041972E-13, f(a,b)(max(a,b))), f(a)(1.0 / sqrt(a))), sum, d0), f(a,b)(a * b)), f(a,b)(a * b)), sum, d0), tensor(d3[1])(1.0), f(a,b)(a * b)), constant(factorization_machine_v1_dense_out_kernel), f(a,b)(a * b)), sum, d3), constant(factorization_machine_v1_dense_out_bias), f(a,b)(a + b)), f(a)(1 / (1 + exp(-a)))), (d1, d2), (d0, d1)), tensor(d0[1])(1.0), f(a,b)(a * b))\n

WARNING: invalid rank feature: 'rankingExpression(firstphase)' (invalid parameters)
ERROR: rank feature verification failed: rankingExpression(firstphase) (first phase ranking)
ERROR: rank profile 'mf_rank': FAIL
```

Diff-ing the files, one can see that the only differences are in the value of 1 constant and the renaming of 2 layers. Changing the names of these layers in the new model from batch_norm/batchnorm/sub/_1__cf__3 -> batch_norm/batchnorm/sub/_1__cf__1 and batch_norm/batchnorm/mul/_0__cf__2 -> batch_norm/batchnorm/mul/_0__cf__0 and I find that the model can be deployed successfully!

In an effort to see whether this was a case of using BatchNorm/Dense layers, I reduced the model to a simple CF model without any additional parameters (see Model_C/factorization_machine_v1/optimized/saved_model.pbtxt https://drive.google.com/open?id=1ONX0iMpro7MDBm-jmgYXxwfePONWfUbc) but obtain a similar type error.

Additionally, I have found similar errors when attempting to deploy to 7.99.22 (our current cluster version).

I have also verified that this problem occurs for TF 1.12.0 and 1.13.1 (haven’t used later versions due to comments read on Gitter and on here re: compatibility).

Furthermore, for the exported SavedModel which worked on 7.74.17 (Model_A/optimized_v2/), if I take the model *before* freezing and optimizing (i.e. with variables.) and deploy to 7.99.22 then I am also able to deploy successfully (assume due to commits related to https://github.com/vespa-engine/vespa/issues/10282).
 
However, taking Model_B which will not work on 7.74.17 without manual renaming, and then deploying the pre-frozen and optimized version (see /factorization_machine_v1/standard/*}  to 7.99.22 this also fails and yields the following error.

```
Invalid application package: default.default: Error loading model: For search cluster 'catalogue', search definition 'product': error in rank setup. Details:
INFO: rank profile 'default': pass
INFO: rank profile 'unranked': pass
ERROR: rank expression contains type errors: join(rename(map(join(reduce(join(join(reduce(join(join(rankingExpression(imported_ml_function_factorization_machine_v1_input_embedding_product_slug), reduce(map(join(join(reduce(map(rankingExpression(imported_ml_function_factorization_machine_v1_input_embedding_product_slug), f(a)(a * a)), sum, d0), tensor(d0[1])(1.0), f(a,b)(a * b)), 9.999999960041972E-13, f(a,b)(max(a,b))), f(a)(1.0 / sqrt(a))), sum, d0), f(a,b)(a * b)), join(rankingExpression(imported_ml_function_factorization_machine_v1_input_embedding_user_guid), reduce(map(join(join(reduce(map(rankingExpression(imported_ml_function_factorization_machine_v1_input_embedding_user_guid), f(a)(a * a)), sum, d0), tensor(d0[1])(1.0), f(a,b)(a * b)), 9.999999960041972E-13, f(a,b)(max(a,b))), f(a)(1.0 / sqrt(a))), sum, d0), f(a,b)(a * b)), f(a,b)(a * b)), sum, d0), tensor(d3[1])(1.0), f(a,b)(a * b)), rename(constant(factorization_machine_v1_dense_out_kernel), (d0, d1), (d3, d2)), f(a,b)(a * b)), sum, d3), rename(constant(factorization_machine_v1_dense_out_bias), d0, d2), f(a,b)(a + b)), f(a)(1 / (1 + exp(-a)))), (d1, d2), (d0, d1)), tensor(d0[1])(1.0), f(a,b)(a * b))\n
WARNING: invalid rank feature: 'rankingExpression(firstphase)' (invalid parameters)
ERROR: rank feature verification failed: rankingExpression(firstphase) (first phase ranking)
ERROR: rank profile 'mf_rank': FAIL
```

Summary: 

Model_A
- optimized can be deployed to Vespa 7.74.17 (see https://drive.google.com/open?id=1lghvIRH5GAN3vDvzLjydeCFkuQ1hmLw7)
- standard can be deployed to Vespa 7.99.22 (unfortunately I do not have this savedmodel.pbtxt to hand)

Model_B
- optimized AND RENAMED can be deployed to Vespa 7.74.17 (see https://drive.google.com/open?id=1ZJ_VMnfl991Jnh4dH7rkruGnJ41tCyHv
- standard cannot be deployed to any 7.74.17 or 7.99.22

Model_C, 
- neither optimized or standard can be deployed to 7.74.17 or 7.99.22 (see https://drive.google.com/open?id=1ONX0iMpro7MDBm-jmgYXxwfePONWfUbc)

Any insight here would be greatly appreciated!

Thanks!

**Comment by lesters** [comment]
@lukealexmiller So, I tried to deploy your models, and identified an issue regarding dimension renaming. I'm working on a fix and should have one ready soon.

Regarding the VarHandleOp, ReadVariableOp etc these were added a few weeks ago, where we started the preparation for supporting TF2.0. We still do have some work there as the underlying SavedModel structure for Keras has changed due to use of the new autograph feature.

**Comment by lukealexmiller** [comment]
Fantastic, thanks for the quick response.

**Comment by lesters** [comment]
@lukealexmiller Thanks for this very detailed report. It took longer than expected to find the issue, but I was able to deploy all your models here after these two:

https://github.com/vespa-engine/vespa/pull/10730
https://github.com/vespa-engine/vespa/pull/10785

The root issue was due to an optimization we have to avoid calculating batch dimensions of size 1. The fix is in Vespa versions >= 7.113.16.

I'll keep the issue open until you are able to verify on your side.

Cheers,
- Lester
